src/corppa/ocr/gvision_ocr.py

```diff
@@ -1,66 +1,293 @@
-                "https://cloud.google.com/apis/design/errors".format(
-                    response.error.message
+            if response.error.message:
+                raise Exception(
+                    f"{response.error.message}\n for more info on error messages, "
+                    "check: https://cloud.google.com/apis/design/errors"
                 )
+
+            # Update counter
+            ocr_count += 1
+            if show_progress:
+                progress_bar.update()
+
+            # Check if we should stop
+            if ocr_limit and ocr_count == ocr_limit:
+                # TODO: Is there a better structuring to avoid this break
+                break
+
+    if show_progress:
+        # Close progress bar
+        progress_bar.close()
+        if ocr_limit and ocr_count == ocr_limit:
+            print("Stopping early, OCR limit reached.", file=sys.stderr)
+        print(
+            f"{ocr_count:,} images OCR'd & {skip_count:,} images skipped.",
+            file=sys.stderr,
+        )
+
+    return {"ocr_count": ocr_count, "skip_count": skip_count}
+
+
+# Should this live somewhere in corppa.utils?
+def get_ppa_source(vol_id):
+    # Note that this is fairly brittle.
+    if vol_id.startswith("CW0") or vol_id.startswith("CB0"):
+        return "Gale"
+    else:
+        return "HathiTrust"
+
+
+def ocr_volumes(vol_ids, in_dir, out_dir, ext_list, ocr_limit=0, show_progress=True):
+    n_vols = len(vol_ids)
+    current_ocr_limit = ocr_limit
+    total_ocr_count = 0
+    total_skip_count = 0
+    for i, vol_id in enumerate(vol_ids):
+        vol_src = get_ppa_source(vol_id)
+        # Currently, this only supports Gale; skip other (i.e. HathiTrust) sources
+        if vol_src != "Gale":
+            print(
+                f"Warning: Skipping volume, since its source '{vol_src}' "
+                "is not yet supported.",
+                file=sys.stderr,
             )
+            continue
+        # Get vol dir info
+        sub_dir = get_vol_dir(vol_src, vol_id)
+        in_vol_dir = in_dir.joinpath(sub_dir)
+        out_vol_dir = out_dir.joinpath(sub_dir)
+
+        # Check that input vol dir exists
+        if not in_vol_dir.is_dir():
+            print(f"Warning: Volume '{vol_id}' is not in {in_dir}", file=sys.stderr)
+            print(f"Directory {in_vol_dir} does not exist.", file=sys.stderr)
+            continue
+        # Ensure that output vol dir exists
+        out_vol_dir.mkdir(parents=True, exist_ok=True)
+        if show_progress:
+            # Add space between volume-level reporting
+            if i:
+                print("", file=sys.stderr)
+            print(f"OCRing {vol_id} ({i+1}/{n_vols})...", file=sys.stderr)
+
+        # OCR images
+        report = ocr_images(
+            in_vol_dir,
+            out_vol_dir,
+            ext_list,
+            ocr_limit=current_ocr_limit,
+            show_progress=show_progress,
+        )
+
```

Comment:
Ah, I see - this is exactly equivalent to the way you were thinking of running it manually, you create the parallel output directories and then run the main `ocr_images` with the input and output directory for this volume. That's not how I'd imagined doing it, which is why I had that comment about potentially pulling out the ocr api chunk of code for reuse (we may still want to do that if it makes unit testing easier). 

It would be nice if we could make a generator of all the images from the specified volumes and have a single progress bar across all the pages form the volumes. Based on how you've structured this code, I think you could make a volume image generator that yields the paths for each volume and makes them relative to the base input directory (not sure yet how you would elegantly configure/pass the generator to the ocr image). I think it would be less code,  but not sure it's worth refactoring at this point unless it simplifies testing.